Command Line Interface Refactor

dr7ana opened the pull request:
Task: Replace current CLI, [cxxopts](https://github.com/jarro2783/cxxopts), with [CLI11](https://github.com/CLIUtils/CLI11)

Fixes #1878

dr7ana committed:
Replace cxxopts with CLI11

- Simiplifies CLI code for future modification
- filesystem library linked in cmake check_for_std_filesystem file

majestrate commented on daemon/lokinet-vpn.cpp:
```diff
@@ -98,119 +104,102 @@ namespace
-    return exit_error(3, "One of --up/--down/--status/--kill must be specified");
-  if (num_commands != 1)
-    return exit_error(3, "Only one of --up/--down/--status/--kill may be specified");
+  switch (options.vpnUp + options.vpnDown + options.printStatus + options.killDaemon)
```
implicit conversion to `int` feels like foot cannon fodder

jagerman replied:
Perhaps keep it explicit via the `int num_commands = ...`

majestrate commented on daemon/lokinet.cpp:
```diff
@@ -382,98 +410,83 @@ lokinet_main(int argc, char** argv)
+  cli.add_option("--config", options.configPath, "Path to lokinet.ini configuration file")
+      ->capture_default_str();
+
 #ifdef _WIN32
```
could probably use:

```
if constexper(llarp:platform::is_windows)
{
   // ...
}
```

instead of a preprocessor macro.

dr7ana replied:
fixed

majestrate commented on daemon/lokinet.cpp:
```diff
@@ -382,98 +410,83 @@ lokinet_main(int argc, char** argv)
       return 0;
     }
+
 #ifdef _WIN32
```
ditto `if constexper` here

jagerman commented on daemon/lokinet-vpn.cpp:
```diff
@@ -1,12 +1,15 @@
 
+#include "CLI/App.hpp"
+#include "CLI/Formatter.hpp"
+#include "CLI/Config.hpp"
```
Should be `<CLI/App.hpp>` etc. for external library includes.

dr7ana replied:
fixed

jagerman commented on daemon/lokinet.cpp:
```diff
@@ -15,11 +15,39 @@
 
+#include "CLI/App.hpp"
+#include "CLI/Formatter.hpp"
+#include "CLI/Config.hpp"
```
`#include <...>`

jagerman commented on daemon/lokinet.cpp:
```diff
@@ -382,98 +410,83 @@ lokinet_main(int argc, char** argv)
-  // clang-format off
-  options.add_options()
+  CLI::App cli{
+      "LokiNET is a free, open source, private, decentralized, market-based sybil resistand and IP "
```
Typo in `resistant`

dr7ana replied:
fixed

jagerman commented on external/CMakeLists.txt:
```diff
@@ -78,7 +78,8 @@ if(WITH_HIVE)
 endif()
 
-add_subdirectory(cxxopts EXCLUDE_FROM_ALL)
+# replacing cxxopts with CLI11
```
Comment seems like a temporary comment to be removed

dr7ana replied:
fixed

jagerman commented on external/readme.md:
```diff
@@ -11,3 +10,4 @@ directory for git submodules
 * pybind11: for pybind modules
 * sqlite_orm: for peer stats db
 * uvw: libuv header only library for main event loop
+* CLI11: cli argument parser
```
Move up to keep this list sorted alphabetically

jagerman committed:
Rename 'package' target on macos to 'dmg' to avoid cpack conflict

CLI11 uses cpack, which makes us fail to configure on macos because of
the 'package' target.  Renaming it to 'dmg' should avoid the conflict.

majestrate commented on external/CMakeLists.txt:
```diff
@@ -78,7 +78,7 @@ if(WITH_HIVE)
 
-add_subdirectory(cxxopts EXCLUDE_FROM_ALL)
+system_or_submodule(CLI11 CLI11 CLI11>=2.2.0 CLI11)
 
```
```
target_link_libraries(CLI11 INTERFACE filesystem)
```
